ci(docs-infra): improve deployment process (reduce manual config steps)

gkalpak opened the pull request:
Previously, the stable branch was always deployed to the `v<X>-angular-io-site` Firebase site, which was connected to the `angular.io` domain. Whenever a new major version was released (and became the new stable version), the `angular.io` domain had to be disconnected from the previous Firebase site and be connected to the new `v<Y>-angular-io-site` Firebase site. This was a manual process that involved making changes in the Firebase console and the DNS records.

This commit is part of a new process that reduces the manual steps as follows:
- A new `stable-angular-io-site` Firebase site will be created.
- The `angular.io` domain will be connected to that new Firebase site.
- When deploying from the stable branch, we will deploy to both `stable-angular-io-site` and `v<X>-angular-io-site`.
- When a new major version is released, we will start deploying from the new stable branch to `stable-angular-io-site`, but there will be no need to connect/disconnect the `angular.io` domain.

---
**TODO**

_**Before merging this PR:**_
- [x] Decide whether we want to keep the redirect from `v<X>.angular.io` to `angular.io` (and whether we want to automate it via Firebase config - currently it happens via Firebase console and/or DNS):
  - Getting rid of the redirect simplifies the setup, but changes the current behavior.
  - Keeping the redirect preserves the current behavior, but requires either manual steps (to set/unset the redirect in Firebase console/DNS) or special Firebase config (to redirect all requests to `angular.io` while stable).
  - Also, all options have slightly different behavior/timing when releasing a new major.

  _Discussed in the `docs-infra` meeting on 2021-03-03 and decided to (a) keep the `v<X>.angular.io` to `angular.io` redirect and (b) look into managing the redirects from `v<X>.angular.io` to other sub-domains via the Firebase config._

_**After merging this PR:**_
- [ ] Wait for the changes to be deployed (both on `master` and the patch branch).
- [ ] Connect `angular.io` and `www.angular.io` to the `stable-angular-io-site` Firebase site.
    (Ensure the domains are disconnected from the `v12-angular-io-site` Firebase site.)
- [ ] Remove the redirect from `v13.angular.io` to `rc.angular.io` (in firebase)
- [ ] Remove the redirect from `v12.angular.io` to `angular.io` (in dns)

mary-poppins commented:
You can preview f061a14 at https://pr40882-f061a14.ngbuilds.io/.

mary-poppins commented:
You can preview a0f9605 at https://pr40882-a0f9605.ngbuilds.io/.

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -118,15 +118,23 @@ function computeDeploymentsInfo(
       deployEnv: 'stable',
       projectId: 'angular-io',
-      siteId: `v${currentBranchMajorVersion}-angular-io-site`,
+      siteId: 'stable-angular-io-site',
```
Do we need this site to be created (and DNS to be setup??) before merging this PR?

gkalpak replied:
Yes, we do need the site (I have already created :grin:), but we should not change the DNS (i.e. connect the `[www.]angular.io` domain to it) until we have a successful deployment (i.e. after the PR has been merged).

petebacondarwin reviewed (approved):
LGTM!

mary-poppins commented:
You can preview 0262349 at https://pr40882-0262349.ngbuilds.io/.

mary-poppins commented:
You can preview ad29dd2 at https://pr40882-ad29dd2.ngbuilds.io/.

mary-poppins commented:
You can preview ac45b9f at https://pr40882-ac45b9f.ngbuilds.io/.

IgorMinar reviewed (approved):
LGTM, thanks for all the tests!

Reviewed-for: docs-infra

mary-poppins commented:
You can preview 889f1df at https://pr40882-889f1df.ngbuilds.io/.

mary-poppins commented:
You can preview 1595179 at https://pr40882-1595179.ngbuilds.io/.

mary-poppins commented:
You can preview 0433dbe at https://pr40882-0433dbe.ngbuilds.io/.

mary-poppins commented:
You can preview 0e479b6 at https://pr40882-0e479b6.ngbuilds.io/.

gkalpak committed:
refactor(docs-infra): move `deploy-to-firebase.js` into a dedicated directory

This commit moves the `aio/scripts/deploy-to-firebase.js` script into a
new directory, `aio/scripts/deploy-to-firebase/`. This will make it
easier to break it up into multiple smaller files to better organize the
code (without polluting the `aio/scripts/` directory).

gkalpak committed:
build(docs-infra): enable linting for `deploy-to-firebase` scripts

This commit enables linting for the scripts in
`aio/scripts/deploy-to-firebase/`. It also makes the necessary changes
to the files to make linting pass.

gkalpak committed:
fix(docs-infra): improve an error message in `deploy-to-firebase` script

This commit fixes a typo in an error message thrown in the
`deploy-to-firebase` script (`but it was but it was` --> `but it was`).

It also makes the error clearer when the request was not redirected.

Old error:
> Expected '<original-url>' to be redirected to '<expected-url>', but it
> was but it was redirected to 'undefined'.

New error:
> Expected '<original-url>' to be redirected to '<expected-url>', but it
> was not redirected.

gkalpak committed:
build(docs-infra): validate computed deployments in `deploy-to-firebase` script

The `deploy-to-firebase` script might have to deploy a single AIO build
to multiple projects/sites (potentially with small tweaks between each).

This commit adds a step to validate the computed deployments to ensure
they are compatible with each other (for example, that there is exactly
one primary deployment that builds the app and sets the theme/mode and
that all secondary deployments are compatible with the primary one).

gkalpak committed:
refactor(docs-infra): convert `deploy-to-firebase` script to ESM

Convert the `deploy-to-firebase` script (and corresponding tests) from
CommonJS format to ESM.

gkalpak committed:
refactor(docs-infra): add `name` and `type` to deployment info

This commit adds two new properties, `name` and `type`, to deployment
info data. This will make it easier to differentiate between deployment
targets (e.g. primary vs secondary) and will provide more informative
log messages.

In a subsequent commit, this new data can be used to validate the
computed deployments list.

gkalpak committed:
refactor(docs-infra): move `deploy-to-firebase` util functions to a separate file

Move some util functions from `deploy-to-firebase/index.mjs` to a
separate file (`deploy-to-firebase/utils.mjs`) to keep `index.mjs` small
and easier to maintain.

NOTE:
To make it easier to mock functions in tests, `utils.mjs` uses a default
export (instead of multiple named exports) and references local
functions through the default exported object.

gkalpak committed:
refactor(docs-infra): create `logSectionHeader()` helper in `deploy-to-firebase` script

Create a `logSectionHeader()` helper function to avoid duplicating code.

gkalpak committed:
refactor(docs-infra): move `deploy-to-firebase` pre-deploy actions to a separate file

Move some functions that are used as pre-deploy actions from
`deploy-to-firebase/index.mjs` to a separate file
(`deploy-to-firebase/pre-deploy-actions.mjs`) to keep `index.mjs` small
and easier to maintain.

Also, rename the `removeServiceWorker()` function to
`disableServiceWorker()`, which is a little more accurate.

NOTE:
While not strictly necessary atm, `pre-deploy-actions.mjs` uses the same
default export pattern for consistency with `utils.mjs`.

gkalpak committed:
refactor(docs-infra): move `deploy-to-firebase` post-deploy actions to a separate file

Move some functions that are used as post-deploy actions from
`deploy-to-firebase/index.mjs` to a separate file
(`deploy-to-firebase/post-deploy-actions.mjs`) to keep `index.mjs` small
and easier to maintain.

NOTE:
While not strictly necessary atm, `post-deploy-actions.mjs` uses the
same default export pattern for consistency with `utils.mjs`.

gkalpak committed:
feat(docs-infra): create undo functions for all pre-deploy actions

Create functions to undo the changes made by any pre-deploy action.
In a future commit, this will allow deploying a build to multiple
projects/sites with different small tweaks for each.

gkalpak committed:
ci: improve angular.io deployment process for the stable branch

Previously, the stable branch was always deployed to the
`v<X>-angular-io-site` Firebase site, which was connected to the
`angular.io` domain. Whenever a new major version was released (and
became the new stable version), the `angular.io` domain had to be
disconnected from the previous Firebase site and be connected to the new
`v<Y>-angular-io-site` Firebase site. This was a manual process that
involved making changes in the Firebase console and the DNS records.

This commit is part of a new process that reduces the manual steps as
follows:
- A new `stable-angular-io-site` Firebase site is created.
- The `angular.io` domain will be connected to that new Firebase site.
- When deploying from the stable branch, we will deploy to both
  `stable-angular-io-site` and `v<X>-angular-io-site`. In addition, the
  deployment to `v<X>-angular-io-site` will update the Firebase config
  file to redirect to `angular.io`.
- When a new major version is released, we will start deploying from the
  new stable branch to `stable-angular-io-site`, but there will be no
  need to connect/disconnect the `angular.io` domain. Also,
  `v<X>.angular.io` will stop redirecting to `angular.io` by means of
  updating the Firebase config file (without requiring changes in the
  Firebase console or DNS).

gkalpak committed:
refactor(docs-infra): make it easier to configure Firebase redirects at deployment

Refactor the `deploy-to-firebase` post-deploy actions by extracting
logic to separate helper functions. This makes it easier to create more
functions for testing various redirect configs (in a future commit).

gkalpak committed:
feat(docs-infra): add more granular support for configuring Firebase redirects at deployment

Previously, there was a `deploy-to-firebase` pre-deploy action for
configuring Firebase to redirect non-file requests to `angular.io`. This
is used for ensuring that `rs.angular.io` is correctly redirected to
`angular.io`, even when people have previously visited (and have a
ServiceWorker activated on) `rc.angular.io`.

This commit adds pre-deploy actions for configuring Firebase to redirect
a deployment to any of `angular.io`, `rc.angular.io` or
`next.angular.io` and also configure whether all requests or only
non-file requests will be redirected.
In a future commit, this will allow managing redirects for the `stable`,
`rc` and `next` deployments via the Firebase config file (without
requiring changes in the Firebase console or DNS).

gkalpak committed:
feat(docs-infra): add post-deploy actions for testing more redirect configs

Add `deploy-to-firebase` post-deploy actions to allow testing various
redirect configs.
In a future commit, this will allow testing redirects for the `stable`,
`rc` and `next` deployments via the Firebase config file (without
requiring changes in the Firebase console or DNS).

gkalpak committed:
ci: improve angular.io deployment process for the RC branch

Previously, (when there was an active RC) the RC branch was only
deployed to the `rc-angular-io-site` Firebase site, which is connected
to the `rc.angular.io` domain. However, if the RC major version was
higher than the stable major version, we also had to manually configure
(via the Firebase console and/or DNS records) the `v<X>.angular.io`
domain to redirect to `rc.angular.io`. Then, once `<X>` became the new
stable version, we had to manually remove the redirect (to let
`v<X>.angular.io` be redirected to `angular.io`).

This commit is part of a new process that reduces the manual steps as
follows (the steps below only apply to RC versions that have a higher
major version than the current stable):
- A `v<X>-angular-io-site` Firebase site will be created for the new RC
  version.
- The `v<X>.angular.io` domain will be connected to that new Firebase
  site.
- When deploying from the RC branch, we will deploy to both
  `rc-angular-io-site` and `v<X>-angular-io-site`. In addition, the
  deployment to `v<X>-angular-io-site` will update the Firebase config
  file to redirect to `rc.angular.io`.
- When the RC version becomes the new stable, we will start deploying to
  `v<X>-angular-io-site` from the stable branch, which will update the
  Firebase config to stop redirecting to `rc.angular.io` and redirect to
  `angular.io` instead (without requiring changes in the Firebase
  console or DNS).

gkalpak committed:
ci: improve angular.io deployment process for the master branch

Previously, the master branch was only deployed to the
`next-angular-io-site` Firebase site, which is connected to the
`next.angular.io` domain. However, if the master major version was
higher than the stable major version (or the RC major version in case
there was an active RC), we also had to manually configure (via the
Firebase console and/or DNS records) the `v<X>.angular.io` domain to
redirect to `next.angular.io`. Then, once `<X>` became the new stable or
RC version, we had to manually remove the redirect (to let
`v<X>.angular.io` be redirected to `angular.io` or `rc.angular.io`).

This commit is part of a new process that reduces the manual steps as
follows (the steps below only apply when the master major version is
higher than the current stable and RC (if applicable)):
- A `v<X>-angular-io-site` Firebase site will be created as soon as the
  version in the `master` branch's `package.json` is updated to a new
  major.
- The `v<X>.angular.io` domain will be connected to that new Firebase
  site.
- When deploying from the master branch, we will deploy to both
  `next-angular-io-site` and `v<X>-angular-io-site`. In addition, the
  deployment to `v<X>-angular-io-site` will update the Firebase config
  file to redirect to `next.angular.io`.
- When the master version becomes the new stable/RC, we will start
  deploying to `v<X>-angular-io-site` from the stable/RC branch, which
  will update the Firebase config to stop redirecting to
  `next.angular.io` and redirect to `(rc.)angular.io` instead (without
  requiring changes in the Firebase console or DNS).

gkalpak committed:
test(docs-infra): avoid running expensive operations when focusing/skipping tests

Previously, the pre-computation of branch version info for the
`deploy-to-firebase` tests would always take place even when the tests
that used that info were skipped. This pre-computation involves network
access and may add a noticeable delay.

This commit refactors the tests to not do the pre-computation if the
corresponding tests are skipped. This allows for a faster iteration
when focusing on a different set of tests.
